sycl/doc/extensions/experimental/sycl_ext_codeplay_cuda_tensor_map.asciidoc

```diff
@@ -0,0 +1,363 @@
+    promote_none,
+    promote_l2_64,
+    promote_l2_128,
+    promote_l2_256,
+  };
+  enum oob_fill : int {
+    oob_fill_none,
+    oob_fill_nan_request_zero_fma,
+  };
+// Implementation defined members must be private
+  im2col_encode_map(queue &q, datatype type, uint32_t rank, void *addr,
+                    const uint64_t gmem_dims[/*rank*/],
+                    const uint64_t gmem_strides[/*rank - 1*/],
+                    const int32_t pixel_box_lower_corner[/*rank*/],
+                    const int32_t pixel_box_upper_corner[/*rank*/],
+                    uint32_t channels_per_pixel, uint32_t pixels_per_col,
+                    const uint32_t element_strides[/*rank*/],
+                    interleave interleave, swizzle swizzle, l2_promote promote,
+                    oob_fill oob_fill);
+};
+}
+----
+
+==== Sample Header for device code only
+
+[source, c++]
+----
+namespace sycl::ext::codeplay::experimental::cuda {
+class tiled_encode_map {
+public:
+  // Get access to the TMA descriptor for use as an operand to the
+  // cp.async.bulk.tensor family of PTX instructions
+  uintptr_t get_native_descriptor();
+};
+class im2col_encode_map {
+  // Get access to the TMA descriptor for use as an operand to the
+  // cp.async.bulk.tensor family of PTX instructions
+ uintptr_t get_native_descriptor();
+};
+}
+----
+
+== Examples
+
+[source, c++]
+----
+#include <cstdint>
+#include <cassert>
+
+#include <sycl/sycl.hpp>
+#include <sycl/ext/codeplay/experimental/cuda_tensor_map.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::codeplay::experimental::cuda;
+#define rank 2
+
+#define WIDTH (256)
+#define HEIGHT (8)
+int main() {
+  device cuda_dev{
+    [](const sycl::device &dev) {
+      return dev.get_backend() == sycl::backend::ext_oneapi_cuda ? 1 : -1;
+    }
+  };
+  bool has_aspect = cuda_dev.has(aspect::ext_codeplay_cuda_tensor_map);
+  assert(has_aspect);
+  queue q{cuda_dev};
+  auto *mem = sycl::malloc_device<int32_t>(WIDTH * HEIGHT, q);
+
+  uint64_t global_dims[rank] = {WIDTH, HEIGHT};
+  uint64_t global_strides[rank - 1] = {WIDTH};
+  uint32_t box_dims[rank] = {WIDTH / 2, HEIGHT / 2};
+  uint32_t element_strides[rank] = {1, 1};
+
+  tiled_encode_map tile(
+    q,
+    static_cast<void *>(mem),
+    tiled_encode_map::datatype::type_int32,
+    rank,
+    global_dims,
+    global_strides,
+    box_dims,
+    element_strides,
+    tiled_encode_map::interleave::interleave_none,
+    tiled_encode_map::swizzle::swizzle_none,
+    tiled_encode_map::l2_promote::promote_none,
+    tiled_encode_map::oob_fill::oob_fill_none
+  );
+
+  q.submit([&](handler &cgh) {
+       local_accessor<int32_t> shmem({WIDTH * HEIGHT}, cgh);
+       local_accessor<int32_t> barrier_mem({WIDTH * HEIGHT}, cgh);
+    cgh.single_task([=]() {
+        auto smem_ptr = reinterpret_cast<uintptr_t>(shmem.get_multi_ptr<access::decorated::yes>().get());
+        auto bar_ptr = reinterpret_cast<uintptr_t>(shmem.get_multi_ptr<access::decorated::yes>().get());
```

Comment:
```suggestion
  q.submit([&](handler &Cgh) {
    sycl_ext::work_group_scratch_size static_size{WIDTH * HEIGHT * sizeof (int32_t)};
    sycl_ext::properties properties{static_size};
    cgh.parallel_for(nd_range<1>(range<1>(Size), range<1>(WgSize)), properties,
                                [=](nd_item<1> Item) {
        sycl_ext::work_group_static<int64_t> barrier_mem;
        auto smem_ptr = reinterpret_cast<uintptr_t>(sycl_ext::static_address_cast<
                      sycl::access::address_space::local_space>(sycl_ext::get_work_group_scratch_memory()).get_decorated());
        auto bar_ptr = reinterpret_cast<uintptr_t>(sycl_ext::static_address_cast<
                      sycl::access::address_space::local_space>(&barrier_mem).get_decorated())
```


Sorry Luke, I missed that before. `local_accessor` usage isn't compatible with the TMA as it will force the compiler to create a copy. This approach will avoid it.

Reply:
> Does this somehow affect affect the kernel ABI?

it is a bit sneaky (note: there is other way to do things, but this is the current behaviour):

To pass a local memory via kernel arg you need to used a specific function to set it (`urSetLocalArg` or something like that), implying you need a top level argument for it. So when the compiler create the actual SPIR-V/PTX/AMDGCN entry point it will add one argument for each `local_accessor` (any accessor actually) data members. We then create a private instance of the functor, initialize the accessor using these arguments and then call `operator()` using this instance as `this`.

in code, the compiler generates something like that (pseudo OpenCL):

```cpp
__kernel void entry(__local char* ptr, int range) {
  sycl_functor obj{};
  obj.acc.__init(ptr, range);
  obj();
}
```

For the operation to work, you need to provide the descriptor address in the `.param` space but, as you probably start to see, we use a private instance of the functor meaning the address will be in SYCL's private address space.

Reply:
for more context, your piece fit with https://github.com/intel/llvm/pull/14332 and https://github.com/intel/llvm/pull/14434

Reply:
It feels to me like a bug that simple, and "obviously correct basic code", would be broken by the way the kernel handles the calling convention of unrelated arguments.


I get that CUDA's ABI requirements here are weird, but I can't help thinking there must be a program-breaking bug in other circumstances.
In any case, I appreciate your example code - and will use it - but I would rather the example code didn't involve several other extensions.

oh well.